File: bot.js
const { Telegraf } = require('telegraf');
const { Octokit } = require('@octokit/rest');
const fs = require('fs').promises;

// Initialize bot and GitHub client
const bot = new Telegraf(process.env.TELEGRAM_BOT_TOKEN);
const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN });

const REPO_OWNER = 'nobuhibiki';
const REPO_NAME = 'nobujournal';
const JOURNAL_FILE = 'journal.json';

async function saveEntry(entry) {
  try {
    // Get current file if it exists
    let currentContent;
    try {
      const { data } = await octokit.repos.getContent({
        owner: REPO_OWNER,
        repo: REPO_NAME,
        path: JOURNAL_FILE,
      });
      currentContent = JSON.parse(Buffer.from(data.content, 'base64').toString());
    } catch (e) {
      currentContent = { entries: [] };
    }

    // Add new entry
    currentContent.entries.push({
      ...entry,
      timestamp: new Date().toISOString(),
    });

    // Update or create file
    await octokit.repos.createOrUpdateFileContents({
      owner: REPO_OWNER,
      repo: REPO_NAME,
      path: JOURNAL_FILE,
      message: 'Add journal entry',
      content: Buffer.from(JSON.stringify(currentContent, null, 2)).toString('base64'),
      sha: currentContent.sha,
    });

    return true;
  } catch (error) {
    console.error('Error saving entry:', error);
    return false;
  }
}

// Command handlers
bot.command('start', async (ctx) => {
  await ctx.reply(
    'Welcome to your GitHub-based Journal Bot! 📝\n\n' +
    'Just send me any message and I\'ll save it to your private GitHub repo.\n' +
    'Use /view to see recent entries\n' +
    'Use /mood [emotion] to log your current mood'
  );
});

bot.command('mood', async (ctx) => {
  const mood = ctx.message.text.split('/mood ')[1];
  if (!mood) {
    await ctx.reply('Please specify your mood: /mood [emotion]');
    return;
  }

  const success = await saveEntry({
    type: 'mood',
    content: mood,
    userId: ctx.from.id,
  });

  if (success) {
    await ctx.reply(`Mood "${mood}" recorded! 🎭`);
  } else {
    await ctx.reply('Sorry, there was an error saving your mood 😕');
  }
});

bot.command('view', async (ctx) => {
  try {
    const { data } = await octokit.repos.getContent({
      owner: REPO_OWNER,
      repo: REPO_NAME,
      path: JOURNAL_FILE,
    });

    const journal = JSON.parse(Buffer.from(data.content, 'base64').toString());
    const userEntries = journal.entries
      .filter(entry => entry.userId === ctx.from.id)
      .slice(-5);

    if (userEntries.length === 0) {
      await ctx.reply('No entries found.');
      return;
    }

    const message = userEntries
      .map(entry => {
        const date = new Date(entry.timestamp).toLocaleString();
        return `📅 ${date}\n${entry.content}\n`;
      })
      .join('\n');

    await ctx.reply(`Your recent entries:\n\n${message}`);
  } catch (error) {
    await ctx.reply('Sorry, there was an error retrieving your entries 😕');
  }
});

// Handle regular messages
bot.on('text', async (ctx) => {
  if (ctx.message.text.startsWith('/')) return;

  const success = await saveEntry({
    type: 'entry',
    content: ctx.message.text,
    userId: ctx.from.id,
  });

  if (success) {
    await ctx.reply('Entry saved! 📝');
  } else {
    await ctx.reply('Sorry, there was an error saving your entry 😕');
  }
});

File: bot.py
from telegram.ext import Application, CommandHandler, MessageHandler, filters
import json
from datetime import datetime
import os

# Load or create journal
def load_journal():
    try:
        with open('journal.json', 'r') as f:
            return json.load(f)
    except FileNotFoundError:
        return {"entries": []}

# Save journal
def save_journal(data):
    with open('journal.json', 'w') as f:
        json.dump(data, f, indent=2)

# Command handlers
async def start(update, context):
    await update.message.reply_text(
        "Welcome to your Journal Bot! 📝\n"
        "Just send me any message to save it as a journal entry.\n"
        "Use /mood [emotion] to log your mood"
    )

async def save_message(update, context):
    journal = load_journal()
    entry = {
        "type": "entry",
        "content": update.message.text,
        "timestamp": datetime.now().isoformat(),
        "user_id": update.message.from_user.id
    }
    journal["entries"].append(entry)
    save_journal(journal)
    await update.message.reply_text("Entry saved! 📝")

async def add_mood(update, context):
    if not context.args:
        await update.message.reply_text("Please specify your mood: /mood [emotion]")
        return
    
    mood = ' '.join(context.args)
    journal = load_journal()
    entry = {
        "type": "mood",
        "content": mood,
        "timestamp": datetime.now().isoformat(),
        "user_id": update.message.from_user.id
    }
    journal["entries"].append(entry)
    save_journal(journal)
    await update.message.reply_text(f"Mood '{mood}' recorded! 🎭")

def main():
    # Get bot token
    token = os.getenv('TELEGRAM_BOT_TOKEN')
    if not token:
        raise ValueError("No token found! Set TELEGRAM_BOT_TOKEN in repository secrets")
    
    # Create application and add handlers
    application = Application.builder().token(token).build()
    
    # Add handlers
    application.add_handler(CommandHandler("start", start))
    application.add_handler(CommandHandler("mood", add_mood))
    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, save_message))
    
    # Run the bot
    application.run_polling()

if __name__ == '__main__':
    main()